tracer/src/Datadog.Trace/Ci/Coverage/DefaultWithGlobalCoverageEventHandler.cs

```diff
@@ -42,108 +40,106 @@ public void Clear()
                     }
+                }
 
-                    List<TypeDef>? moduleTypes;
-                    lock (TypeDefsFromModuleDefs)
+                var componentCoverageInfos = new Dictionary<Module, ComponentCoverageInfo>();
+                var fileCoverageInfos = new Dictionary<FileCoverageMetadata, FileCoverageInfo>();
+
+                var fileBitmapBuffer = stackalloc byte[512];
+                foreach (var moduleValue in GetModuleValues())
+                {
+                    var module = moduleValue.Module;
+                    if (!componentCoverageInfos.TryGetValue(module, out var componentCoverageInfo))
                     {
-                        if (!TypeDefsFromModuleDefs.TryGetValue(moduleDef, out moduleTypes))
-                        {
-                            moduleTypes = moduleDef.GetTypes().ToList();
-                            TypeDefsFromModuleDefs[moduleDef] = moduleTypes;
-                        }
+                        componentCoverageInfo = new ComponentCoverageInfo(module.Name);
+                        globalCoverage.Components.Add(componentCoverageInfo);
+                        componentCoverageInfos[module] = componentCoverageInfo;
                     }
 
-                    var componentCoverageInfo = new ComponentCoverageInfo(moduleDef.FullName);
-                    for (var mIdx = 0; mIdx < moduleValue.Metadata.GetMethodsCount(); mIdx++)
+                    foreach (var moduleFile in moduleValue.Metadata.Files)
                     {
-                        var methodValue = moduleValue.Methods[mIdx];
-                        if (methodValue is null && moduleProcessed.Contains(moduleValue.Module))
+                        if (!fileCoverageInfos.TryGetValue(moduleFile, out var fileCoverageInfo))
                         {
-                            continue;
-                        }
+                            fileCoverageInfo = new FileCoverageInfo(moduleFile.Path)
+                            {
+                                ExecutableBitmap = moduleFile.Bitmap
+                            };
 
-                        moduleValue.Metadata.GetMethodsMetadata(mIdx, out var typeIndex, out var methodIndex);
-                        var typeDef = moduleTypes[typeIndex];
-                        var methodDef = typeDef.Methods[methodIndex];
+                            componentCoverageInfo.Files.Add(fileCoverageInfo);
+                            fileCoverageInfos[moduleFile] = fileCoverageInfo;
+                        }
 
-                        if (methodDef.HasBody && methodDef.Body.HasInstructions)
+                        var fileBitmapLastExecutableLine = moduleFile.LastExecutableLine;
+                        var fileBitmapSize = FileBitmap.GetSize(fileBitmapLastExecutableLine);
+                        using var fileBitmap = fileBitmapSize <= 512 ? new FileBitmap(fileBitmapBuffer, fileBitmapSize) : new FileBitmap(new byte[fileBitmapSize]);
```

Comment:
In this kind of situation, I'd really prefer if you could use ref structs to ensure that the instance doesn't outlive the stackallocated buffer, or at the very least wrap all the usage of the FileBitmap in a method (to reduce the probability that somebody updates the code in such a way that it extends the lifetime of the FileBitmap)